File: gnome-sort/gnome-sort.ts
const swap = (list: number[], i1: number, i2: number): number[] => {
  const tmp = list.slice();
  [tmp[i1], tmp[i2]] = [tmp[i2], tmp[i1]];
  return tmp;
};

type GnomeSort = (list: number[], pointer?: number) => number[];
const gnomeSort: GnomeSort = (list, pointer = 0) => {
  const listSize = list.length;
  const prevPointer = pointer - 1;
  const nextPointer = pointer + 1;

  if (listSize <= 1) return list;

  if (pointer > listSize - 1 || pointer < 0) throw Error();
  if (pointer === 0) return gnomeSort(list, nextPointer);
  if (pointer === listSize - 1) return list;

  if (list[pointer] < list[prevPointer])
    return gnomeSort(swap(list, pointer, prevPointer), prevPointer);
  return gnomeSort(list, nextPointer);
};

const arr = Object.freeze([23, 73, 1, 3, 71, 10, 16, 63, 125, 52, 156, 22]);
const res = gnomeSort(arr.slice(), 1);

console.log("Dataset:", arr);
console.log("Result:", res);

File: gnome-sort/gnome-sort.c
#include <stdio.h>

int list[] = {2, 56, 66, 51, 82, 512, 92, 3, 8, 5, 8, 0, 11, 36, 437};
void gnomeSort(int l[], int s);
void printArr(int arr[], size_t arrLength);

int main() {
  int listSize = sizeof(list) / sizeof(list[0]);
  gnomeSort(list, listSize);
  printArr(list, listSize);
  return 0;
}

void gnomeSort(int l[], int s) {
  if (s < 2)
    return;
  int p = 1;
  while (p < s)
    if (p == 0)
      ++p;
    else if (l[p] < l[p - 1]) {
      int t = l[p];
      l[p] = l[p - 1];
      l[p - 1] = t;
      --p;
    } else
      ++p;
}

void printArr(int arr[], size_t arrLength) {
  for (int x = 0; x < arrLength; x++)
    printf("%i, ", arr[x]);
  printf("\n");
}